// @atlaskit/motion/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ExitingPersistence", {
  enumerable: true,
  get: function get() {
    return _exitingPersistence.default;
  }
});
Object.defineProperty(exports, "FadeIn", {
  enumerable: true,
  get: function get() {
    return _fadeIn.default;
  }
});
Object.defineProperty(exports, "ResizingHeight", {
  enumerable: true,
  get: function get() {
    return _height.ResizingHeight;
  }
});
Object.defineProperty(exports, "ShrinkOut", {
  enumerable: true,
  get: function get() {
    return _shrinkOut.default;
  }
});
Object.defineProperty(exports, "SlideIn", {
  enumerable: true,
  get: function get() {
    return _slideIn.default;
  }
});
Object.defineProperty(exports, "StaggeredEntrance", {
  enumerable: true,
  get: function get() {
    return _staggeredEntrance.default;
  }
});
Object.defineProperty(exports, "ZoomIn", {
  enumerable: true,
  get: function get() {
    return _zoomIn.default;
  }
});
Object.defineProperty(exports, "durationStep", {
  enumerable: true,
  get: function get() {
    return _durations.durationStep;
  }
});
Object.defineProperty(exports, "durations", {
  enumerable: true,
  get: function get() {
    return _durations.durations;
  }
});
Object.defineProperty(exports, "easeIn", {
  enumerable: true,
  get: function get() {
    return _curves.easeIn;
  }
});
Object.defineProperty(exports, "easeInOut", {
  enumerable: true,
  get: function get() {
    return _curves.easeInOut;
  }
});
Object.defineProperty(exports, "easeOut", {
  enumerable: true,
  get: function get() {
    return _curves.easeOut;
  }
});
Object.defineProperty(exports, "exitingDurations", {
  enumerable: true,
  get: function get() {
    return _durations.exitingDurations;
  }
});
Object.defineProperty(exports, "isReducedMotion", {
  enumerable: true,
  get: function get() {
    return _accessibility.isReducedMotion;
  }
});
Object.defineProperty(exports, "linear", {
  enumerable: true,
  get: function get() {
    return _curves.linear;
  }
});
Object.defineProperty(exports, "prefersReducedMotion", {
  enumerable: true,
  get: function get() {
    return _accessibility.prefersReducedMotion;
  }
});
Object.defineProperty(exports, "reduceMotionAsPerUserPreference", {
  enumerable: true,
  get: function get() {
    return _accessibility.reduceMotionAsPerUserPreference;
  }
});
Object.defineProperty(exports, "useExitingPersistence", {
  enumerable: true,
  get: function get() {
    return _exitingPersistence.useExitingPersistence;
  }
});
Object.defineProperty(exports, "useIsReducedMotion", {
  enumerable: true,
  get: function get() {
    return _accessibility.useIsReducedMotion;
  }
});
Object.defineProperty(exports, "useResizingHeight", {
  enumerable: true,
  get: function get() {
    return _height.useResizingHeight;
  }
});
var _durations = require("./utils/durations");
var _curves = require("./utils/curves");
var _accessibility = require("./utils/accessibility");
var _fadeIn = _interopRequireDefault(require("./entering/fade-in"));
var _staggeredEntrance = _interopRequireDefault(require("./entering/staggered-entrance"));
var _height = require("./resizing/height");
var _exitingPersistence = _interopRequireWildcard(require("./entering/exiting-persistence"));
var _zoomIn = _interopRequireDefault(require("./entering/zoom-in"));
var _slideIn = _interopRequireDefault(require("./entering/slide-in"));
var _shrinkOut = _interopRequireDefault(require("./entering/shrink-out"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
