// @atlaskit/theme/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CHANNEL", {
  enumerable: true,
  get: function get() {
    return _constants.CHANNEL;
  }
});
Object.defineProperty(exports, "DEFAULT_THEME_MODE", {
  enumerable: true,
  get: function get() {
    return _constants.DEFAULT_THEME_MODE;
  }
});
Object.defineProperty(exports, "THEME_MODES", {
  enumerable: true,
  get: function get() {
    return _constants.THEME_MODES;
  }
});
Object.defineProperty(exports, "assistive", {
  enumerable: true,
  get: function get() {
    return _constants.assistive;
  }
});
Object.defineProperty(exports, "borderRadius", {
  enumerable: true,
  get: function get() {
    return _constants.borderRadius;
  }
});
exports.colors = exports.colorPalettes = void 0;
Object.defineProperty(exports, "createTheme", {
  enumerable: true,
  get: function get() {
    return _createTheme.createTheme;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _theme.default;
  }
});
Object.defineProperty(exports, "focusRing", {
  enumerable: true,
  get: function get() {
    return _constants.focusRing;
  }
});
Object.defineProperty(exports, "getTheme", {
  enumerable: true,
  get: function get() {
    return _getTheme.default;
  }
});
Object.defineProperty(exports, "gridSize", {
  enumerable: true,
  get: function get() {
    return _constants.gridSize;
  }
});
Object.defineProperty(exports, "layers", {
  enumerable: true,
  get: function get() {
    return _constants.layers;
  }
});
Object.defineProperty(exports, "noFocusRing", {
  enumerable: true,
  get: function get() {
    return _constants.noFocusRing;
  }
});
Object.defineProperty(exports, "skeletonShimmer", {
  enumerable: true,
  get: function get() {
    return _constants.skeletonShimmer;
  }
});
Object.defineProperty(exports, "themed", {
  enumerable: true,
  get: function get() {
    return _themed.default;
  }
});
exports.typography = void 0;
Object.defineProperty(exports, "useGlobalTheme", {
  enumerable: true,
  get: function get() {
    return _theme.useGlobalTheme;
  }
});
Object.defineProperty(exports, "visuallyHidden", {
  enumerable: true,
  get: function get() {
    return _constants.visuallyHidden;
  }
});
var colorPalettes = _interopRequireWildcard(require("./color-palettes"));
exports.colorPalettes = colorPalettes;
var colors = _interopRequireWildcard(require("./colors"));
exports.colors = colors;
var typography = _interopRequireWildcard(require("./typography"));
exports.typography = typography;
var _getTheme = _interopRequireDefault(require("./utils/get-theme"));
var _themed = _interopRequireDefault(require("./utils/themed"));
var _constants = require("./constants");
var _theme = _interopRequireWildcard(require("./components/theme"));
var _createTheme = require("./utils/create-theme");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
